README.md
# SmartObjects JavaScript Client

[![Build status](https://travis-ci.org/mnubo/smartobjects-js-client.svg?branch=master)](https://travis-ci.org/mnubo/smartobjects-js-client)
[![npm version](https://badge.fury.io/js/mnubo-sdk.svg)](https://www.npmjs.com/package/mnubo-sdk)

## Quickstart

[comment]: # (Important: leave the HTML in this section)
[comment]: # (quickstart-setup)

<h3>Getting the client library</h3>
<p>The client library is available on <a target="_blank" href="https://www.npmjs.com/package/mnubo-sdk">NPM</a>.</p>

<p>Below is an example of how you can install everything:</p>
<pre>
    <code>
npm install --save es6-shim  # if you run node < 4.0.0
npm install --save mnubo-sdk
    </code>
</pre>

<p>For more information, visit <a target="_blank" href="https://github.com/mnubo/smartobjects-js-client">GitHub</a>.<p>

<h3>Create a client instance</h3>

<p>The following JavaScript code can be used to create an instance:</p>

<pre>
    <code>
var mnubo = require('mnubo-sdk');

/* Create a new client with client id and client secret. */
var client = new mnubo.Client({
  id: '<%= clientKey %>',
  secret: '<%= clientSecret %>',
  httpOptions = {
    protocol: 'https',
    hostname: "<%= hostname %>",
    port: 443
  };
});
    </code>
</pre>

[comment]: # (quickstart-setup)

## Introduction

This is a JavaScript opinionated version of the original [API documentation](https://smartobjects.mnubo.com/documentation/). Use classes Owners, Objects, Events, and Search to manage your owners, objects, events and perform searches using the API.

## Prerequisites

The current SDK version only works in a NodeJS environment. For security reasons, it is not recommended to use this SDK in a browser environment because the client id and secret could easily be discovered.

The minimum requirement is `node>=0.10.40` but because the code uses [Promise](https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/Promise.jsm/Promise) and [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map), if you are not running `node>=4.0.0`, you need to require `es6-shim` or any other library that polyfills those structures.

    npm install --save es6-shim

    require('es6-shim'); // before loading mnubo-sdk

## Installation

    npm install --save mnubo-sdk

## Usage

Because the platform supports UTF-8 for ingesting data, the encoding to use in the client must be UTF-8 as well.

### Authentication

The authentication is wrapped for every SDK call. The library will first fetch a new Access Token and make the API call. There is nothing to do from a developer's perspective besides setting the client id, client secret and environment during initialization.

### Initialization

The initialization of the SDK client requires two mandatory fields:

- `id` (**mandatory**): The client id.
- `secret` (**mandatory**): The client secret.
- `env` (*optional*): The environment where the API calls will be sent. It can be either `sandbox` or `production`. By default the `env` is `sandbox`.
- `compression` (*optional*): If you want to use `gzip` compression for both inbound and outbound data, set the value to `true`. By default this parameter is `false`.
- `exponentialBackoff` (*optional*): If you want to use exponential back off retries. By default, the feature is off (`null`).

```
/* Load mnubo SDK. */
require('es6-shim'); /* only if running node < 4.0.0 */
var mnubo = require('mnubo-sdk');

/* Create a new client with client id and client secret. */
var client = new mnubo.Client({
  id: 'KyR2LipTvH2ltAML4ScOTZ7TPLZvjV6oVYIYyx6CZORsoa131d',
  secret: 'B0KcvAeSZJuSXg9Hi0IYZKcsCNNUkw8bxsZ0GSgAW5cWKARa6m',
  env: 'production'
});
```

You can generate an App Token from the web application and use it with this SDK.

Note: App Tokens can have restricted access to specific APIs and can be disabled by an administrator at any given time. This means that some API provided in this SDK may not work correctly.

```js
/* Create a new client with an access token. */
var client = new mnubo.Client({
  token: 'token...',
  env: 'sandbox'
});
```

#### Compression
You can also fine grain the `compression` by using this syntax:

```
compression: {
  requests: true,  /* sets Accept-Encoding: gzip */
  responses: true, /* sets Content-Encoding: gzip */
}
```

Note that, when using compression, the data being resolved in the Promise will always be JSON. The compression/decompression is done behind the scenes using NodeJS `zlib` library.

#### Exponential backoff retries
Exponential backoff retries allows a user of the client to automatically retry requests when a `503` HTTP error is returned from the `SmartObjects` platform. It will gradually increase the interval between retries to allow the server to recover. The configuration to turn on the feature looks like this:

```typescript
exponentialBackoff: {
  numberOfAttempts: 2, /* How many retries to make before failing. Default to 5 if undefined. */
  initialDelayInMillis: 100, /* The number of ms to wait before the first retry. Default to 500 if undefined. */
  onRetry: (attempt: number) => { /* A callback function called for every retry. Default to nothing if undefined. */
    console.log('Attempt #' + attempt);
  }
}
```

### API Calls

All the API calls return a [Promise](https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/Promise.jsm/Promise).

- When a promise is successful, you can call the `.then()` function to get the data returned by the mnubo servers. If there is no data, the value is `null`. ex: `client.events.send({...}).then(function(data) { console.log(data); });`

- When a promise has failed, you can call the `.catch()` function to get the error returned by the mnubo servers. If there is no data, the value is `null`. ex: `client.events.send({...}).catch(function(error) { console.log(error); });`

If you are not familiar with promises, there is an excellent article on [html5rocks](http://www.html5rocks.com/en/tutorials/es6/promises/).

### Examples

If you need some examples to get started with the SDK, you can check out the [wiki page](https://github.com/mnubo/smartobjects-js-client/wiki/Examples).

## Development

With Visual Studio code, you can use the development container extension. This will open
the editor in a container that has all the requirement while leaving your workstation
untouched.

From the editor, you can then, open a terminal and do the following to run the tests:
```bash
root@4d7a461e5fbc:/workspaces/smartobjects-js-client# source scripts/test-setup.sh YOUR_KEY YOUR_SECRET
root@4d7a461e5fbc:/workspaces/smartobjects-js-client# npm install
root@4d7a461e5fbc:/workspaces/smartobjects-js-client# npm run build
root@4d7a461e5fbc:/workspaces/smartobjects-js-client# npm test
```

## References

[mnubo documentation](https://smartobjects.mnubo.com/documentation/)

[Promise](https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/Promise.jsm/Promise)

[Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)

[Examples](https://github.com/mnubo/smartobjects-js-client/wiki/Examples)


CHANGELOG.md
## [1.12.1](https://github.com/mnubo/smartobjects-js-client/compare/1.12.0...1.12.1) (2020-01-24)


### Bug Fixes

* **types:** id secret are optional if token is given ([4eb0b84](https://github.com/mnubo/smartobjects-js-client/commit/4eb0b84))



# [1.12.0](https://github.com/mnubo/smartobjects-js-client/compare/1.11.0...1.12.0) (2019-06-13)


### Bug Fixes

* replace New Buffer with Buffer.from ([64113e8](https://github.com/mnubo/smartobjects-js-client/commit/64113e8))
* **client:** return chained promise to avoid unhandled promise rejections ([1fb8638](https://github.com/mnubo/smartobjects-js-client/commit/1fb8638)), closes [#19](https://github.com/mnubo/smartobjects-js-client/issues/19)


### Features

* **model:** add/remove relation to types ([c1996f4](https://github.com/mnubo/smartobjects-js-client/commit/c1996f4))
* **setup:** add setup template ([5458f34](https://github.com/mnubo/smartobjects-js-client/commit/5458f34))



<a name="1.11.1"></a>
## [1.11.1](https://github.com/mnubo/smartobjects-js-client/compare/1.11.0...1.11.1) (2018-10-12)


### Bug Fixes

* replace New Buffer with Buffer.from ([64113e8](https://github.com/mnubo/smartobjects-js-client/commit/64113e8))
* **client:** return chained promise to avoid unhandled promise rejections ([1fb8638](https://github.com/mnubo/smartobjects-js-client/commit/1fb8638)), closes [#19](https://github.com/mnubo/smartobjects-js-client/issues/19)


### Features

* **setup:** add setup template ([5458f34](https://github.com/mnubo/smartobjects-js-client/commit/5458f34))



<a name="1.11.0"></a>
# [1.11.0](https://github.com/mnubo/smartobjects-js-client/compare/1.10.1...1.11.0) (2018-01-24)


### Features

* **modeler:** add the modeler API along with tests ([2f57f32](https://github.com/mnubo/smartobjects-js-client/commit/2f57f32))
* **version:** set the X-MNUBO-SDK version in a header ([ec1a16f](https://github.com/mnubo/smartobjects-js-client/commit/ec1a16f))



<a name="1.10.1"></a>
## [1.10.1](https://github.com/mnubo/smartobjects-js-client/compare/1.10.0...1.10.1) (2017-12-21)


### Bug Fixes

* **http:** convert payload to Buffer to get proper utf-8 encoding ([35ecbe3](https://github.com/mnubo/smartobjects-js-client/commit/35ecbe3)), closes [#12](https://github.com/mnubo/smartobjects-js-client/issues/12)



<a name="1.10.0"></a>
# [1.10.0](https://github.com/mnubo/smartobjects-js-client/compare/1.9.0...1.10.0) (2017-08-18)


### Features

* **token:** use a token instead of client/secret ([f15813e](https://github.com/mnubo/smartobjects-js-client/commit/f15813e))



<a name="1.9.0"></a>
# [1.9.0](https://github.com/mnubo/smartobjects-js-client/compare/1.8.1...1.9.0) (2017-07-12)


### Bug Fixes

* **test:** fix flaky test and format it ([46d386a](https://github.com/mnubo/smartobjects-js-client/commit/46d386a))


### Features

* **retry:** support exponential back off retry ([c994344](https://github.com/mnubo/smartobjects-js-client/commit/c994344))
* **types:** add type definition for promise-retry ([29f7102](https://github.com/mnubo/smartobjects-js-client/commit/29f7102))



<a name="1.8.1"></a>
## [1.8.1](https://github.com/mnubo/smartobjects-js-client/compare/1.8.0...1.8.1) (2017-05-01)


### Bug Fixes

* **authentication:** remove Accept-Encoding header ([7b77f74](https://github.com/mnubo/smartobjects-js-client/commit/7b77f74))



<a name="1.8.0"></a>
# [1.8.0](https://github.com/mnubo/smartobjects-js-client/compare/1.7.0...1.8.0) (2017-04-28)


### Features

* **model:** add export model functionality ([e766108](https://github.com/mnubo/smartobjects-js-client/commit/e766108))
* **owners:** support a body for claim/unclaim of objects ([1768933](https://github.com/mnubo/smartobjects-js-client/commit/1768933))



<a name="1.7.0"></a>
# [1.7.0](https://github.com/mnubo/smartobjects-js-client/compare/1.5.0...1.7.0) (2017-04-03)


### Features

* **bigdata:** add lowlevel api for bigdata ([9c6159e](https://github.com/mnubo/smartobjects-js-client/commit/9c6159e))
* **events:** add options for send from device ([42f29c6](https://github.com/mnubo/smartobjects-js-client/commit/42f29c6))



<a name="1.6.0"></a>
# [1.6.0](https://github.com/mnubo/smartobjects-js-client/compare/1.5.0...1.6.0) (2017-03-17)


### Features

* **owners:** add batch claim and unclaim ([222cc75](https://github.com/mnubo/smartobjects-js-client/commit/222cc75))



<a name="1.5.0"></a>
# [1.5.0](https://github.com/mnubo/smartobjects-js-client/compare/1.4.0...1.5.0) (2016-12-07)


### Bug Fixes

* **compression:** properly handle types and string conversion ([dcb1297](https://github.com/mnubo/smartobjects-js-client/commit/dcb1297))
* **mnubo:** make ClientOptions.env optional ([007cbe6](https://github.com/mnubo/smartobjects-js-client/commit/007cbe6))


### Code Refactoring

* remove window related files ([b515bed](https://github.com/mnubo/smartobjects-js-client/commit/b515bed))


### Features

* **owners:** add unclaim of deviceId ([3717112](https://github.com/mnubo/smartobjects-js-client/commit/3717112))


### BREAKING CHANGES

* remove any reference to mnubo.static and other related `window` object.
    Now you can only use the `node` related files and objects.



<a name="1.4.0"></a>
# [1.4.0](https://github.com/mnubo/smartobjects-js-client/compare/1.3.0...1.4.0) (2016-09-01)


### Features

* **compression:** introduce gzip compression to requests and responses ([a072fc9](https://github.com/mnubo/smartobjects-js-client/commit/a072fc9))



<a name="1.3.0"></a>
# [1.3.0](https://github.com/mnubo/smartobjects-js-client/compare/1.2.0...1.3.0) (2016-08-03)


### Features

* **events:** allow `reportResults` and `objectsMustExist` options in send ([993111a](https://github.com/mnubo/smartobjects-js-client/commit/993111a))
* **events:** introduce exists APIs for events ([c378fe2](https://github.com/mnubo/smartobjects-js-client/commit/c378fe2))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/mnubo/smartobjects-js-client/compare/1.1.0...1.2.0) (2016-07-12)


### Bug Fixes

src/mnubo.ts
import { base64Encode } from './utils/underscore';
import { http } from './http/http';
import { RequestOptions } from './http/request';
import * as promiseRetry from 'promise-retry';

import { Owners } from './ingestion/owners';
import { Objects } from './ingestion/objects';
import { Events } from './ingestion/events';
import { Search } from './restitution/search';
import { Bigdata } from './bigdata/bigdata';
import { Model } from './model/model';

let version: string;
try {
  version = require('../package.json').version;
} catch (e) {
  version = 'unknown';
}

export enum OAuth2Scopes {
  ALL,
  LIMITED,
  READ,
  WRITE,
}

export interface ClientCompression {
  requests: boolean;
  responses: boolean;
}

export interface ExponentialBackoff {
  numberOfAttempts?: number;
  initialDelayInMillis?: number;
  onRetry?: (attempt: number) => any;
}

export interface ClientOptions {
  id?: string;
  secret?: string;
  env?: string;
  token?: string;
  httpOptions?: RequestOptions;
  compression?: boolean | ClientCompression;
  exponentialBackoff?: ExponentialBackoff;
}

function isClientCompression(object: boolean | ClientCompression): object is ClientCompression {
  return (<ClientCompression>object).requests !== undefined || (<ClientCompression>object).responses !== undefined;
}

class AccessToken {
  private requestedAt: Date;

  constructor(public value: string, public type: string, public expiresIn: number, public jti: string) {
    this.requestedAt = new Date();
  }

  isValid() {
    const now = new Date();

    return now.getTime() < this.requestedAt.getTime() + this.expiresIn * 1000;
  }
}

export class Client {
  private defaultNumberOfAttempts: number = 5;
  private defaultInitialDelay: number = 500;

  public owners: Owners;
  public objects: Objects;
  public events: Events;
  public search: Search;
  public bigdata: Bigdata;
  public model: Model;

  private token: AccessToken;

  constructor(public options: ClientOptions) {
    if (options.env === undefined) {
      options.env = 'sandbox';
    }

    if (!options.httpOptions) {
      options.httpOptions = {
        protocol: 'https',
        hostname: this.hostname(),
        port: 443,
      };
    }

    if (!options.compression) {
      options.compression = false;
    }

    this.owners = new Owners(this);
    this.objects = new Objects(this);
    this.events = new Events(this);
    this.search = new Search(this);
    this.bigdata = new Bigdata(this);
    this.model = new Model(this);
  }

  hostname(): string {
    let part = 'sandbox';

    if (this.options.env === 'production') {
      part = 'api';
    }

    return `rest.${part}.mnubo.com`;
  }

  getAccessToken(scope?: OAuth2Scopes): Promise<any> {
    scope = scope || OAuth2Scopes.ALL;

    const id = this.options.id;
    const secret = this.options.secret;
    const payload = `grant_type=client_credentials&scope=${OAuth2Scopes[scope].toUpperCase()}`;

    const options: RequestOptions = {
      path: '/oauth/token',
      headers: new Map<string, string>(),
    };

    options.headers.set('Authorization', `Basic ${base64Encode(id + ':' + secret)}`);
    options.headers.set('Content-Type', 'application/x-www-form-urlencoded');
    options.headers.set('X-MNUBO-SDK', `JavaScript/${version}`);

    Object.assign(options, this.options.httpOptions);

    return http.post(options, payload).then((data: any) => {
      this.token = new AccessToken(data.access_token, data.token_type, data.expires_in, data.jti);
    });
  }

  /**
   * Is the access token still valid?

   * @return {boolean} false if there is no access token or if it has expired.
   */
  isAccessTokenValid(): boolean {
    return this.token && this.token.isValid();
  }

  authenticate(): Promise<any> {
    if (this.isAccessTokenValid()) {
      return Promise.resolve(this.token);
    } else {
      return this.getAccessToken().then(() => {
        return this.token;
      });
    }
  }

  buildHttpOptions(path: string, contentType?: string): RequestOptions {
    const options: RequestOptions = {
      path: path,
      headers: new Map<string, string>(),
    };
    const compressionAlgorithm = 'gzip';
    const compression = this.options.compression;

    contentType = contentType || 'application/json';

    const token = this.options.token ? this.options.token : this.token.value;

    options.headers.set('Authorization', `Bearer ${token}`);
    options.headers.set('Content-Type', `${contentType}`);

    if (isClientCompression(compression)) {
      if (compression.requests === true) {
        options.headers.set('Content-Encoding', compressionAlgorithm);
      }
      if (compression.responses === true) {
        options.headers.set('Accept-Encoding', compressionAlgorithm);
      }
    } else if (compression === true) {
      options.headers.set('Accept-Encoding', compressionAlgorithm);
      options.headers.set('Content-Encoding', compressionAlgorithm);
    }

    Object.assign(options, this.options.httpOptions);

    return options;
  }

  possiblyRetry(f: () => Promise<any>): Promise<any> {
    const exponentialBackoff = this.options.exponentialBackoff;
    if (!exponentialBackoff) {
      return f().catch((err) => Promise.reject(err.payload));
    } else {
      const retries = exponentialBackoff.numberOfAttempts
        ? exponentialBackoff.numberOfAttempts
        : this.defaultNumberOfAttempts;
      const minTimeout = exponentialBackoff.initialDelayInMillis
        ? exponentialBackoff.initialDelayInMillis
        : this.defaultInitialDelay;

      const opts: any = {
        retries,
        minTimeout,
        randomize: true,
      };

      return promiseRetry((retry: any, attempt: number) => {
        return f().catch((err) => {
          if (err.statusCode === 503) {
            if (exponentialBackoff.onRetry && attempt <= exponentialBackoff.numberOfAttempts) {
              exponentialBackoff.onRetry(attempt);
            }
            return retry();
          } else {
            return Promise.reject(err.payload);
          }
        });
      }, opts);
    }
  }

  get(path: string, contentType?: string) {
    return this.possiblyRetry(() => http.get(this.buildHttpOptions(path, contentType)));
  }

  post(path: string, payload: any, contentType?: string) {
    return this.possiblyRetry(() => http.post(this.buildHttpOptions(path, contentType), payload));
  }

  put(path: string, payload: any, contentType?: string) {
    return this.possiblyRetry(() => http.put(this.buildHttpOptions(path, contentType), payload));
  }

  delete(path: string, contentType?: string) {
    return this.possiblyRetry(() => http.delete(this.buildHttpOptions(path, contentType)));
  }
}
